src/ws/session-utils.c

```diff
@@ -399,11 +399,61 @@ open_session (pam_handle_t *pamh)
   return PAM_SUCCESS;
 }
 
+static void
+do_fd_remapping (int *remap_fds, int n_remap_fds)
+{
+  /* we need to get all of the remap-fds to be numerically above
+   * n_remap_fds in order to make sure that we don't overwrite them in
+   * the middle of the dup2() loop below, and also avoid the case that
+   * dup2() is a no-op (which could fail to clear the O_CLOEXEC flag,
+   * for example).
+   *
+   * we're sloppy with the old fd values because we know that the fdwalk
+   * below will clean everything up.
+   */
+  for (int i = 0; i < n_remap_fds; i++)
+    if (remap_fds[i] != -1) /* -1 == no-op */
+      {
+        /* keep trying until we get a large enough fd.
+         *
+         * it might already be large enough, in which case we don't have
+         * to do anything.
+         */
+        while (remap_fds[i] < n_remap_fds)
+          {
+            int new_fd = dup (remap_fds[i]);
```

Comment:
There must be a better way to do this than repeatedly duping an fd until it is large enough. This really feels unnecessarily complicated. Couldn't it just start at `n_remap_fds` right away and use dup2()? You are going to close all higher fds anyway, so if that kills an existing one it doesn't matter, right?

Reply:
`fcntl(F_DUPFD)`.